### PR TITLE
Sarah adelaide/unit tests

### PR DESCRIPTION
**Pull request by SarahAdelaide**
Added some simple unit tests to alarms and for the parameter config bounding

### CODE REVIEW

File: controller/test/alarm/alarm_test.cpp [hunk 1]
```diff
@@ -20,10 +20,45 @@ TEST_F(AlarmTest, AddOneAlarm) {
   ASSERT_TRUE(alarm_available());
 }
 
+TEST_F(AlarmTest, RemoveOneAlarm) {
+  alarm_remove();
```
Comment (jlebar) [inline controller/test/alarm/alarm_test.cpp:24]: Is the intent of this test to check that we can remove an alarm after having added one alarm (in `alarm_add`)?  Or is the intent to check that alarm_remove() is OK to call if there are no alarms present?

Right now it's doing the former, but that's relying on tests being run in order, and on nobody modifying AddOneAlarm.  To do the latter, we'd need to reolve the TODO in `SetUp`.

File: controller/test/parameters/parameters_test.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
+#include "parameters.h"
+#include "gtest/gtest.h"
+
+TEST(Parameters, SetAndGetRR) {
+  parameters_setRR(RR_MAX + 1);
+  float rr = parameters_getRR();
+  ASSERT_FLOAT_EQ(RR_MAX, rr);
```
Comment (jlebar) [inline controller/test/parameters/parameters_test.cpp:7]: Nit, one-liner?

`ASSERT_FLOAT_EQ(RR_MAX, parameters_getRR())`

Same below.

File: controller/test/alarm/alarm_test.cpp (hunk 1 above)
Comment (SarahAdelaide) [reply controller/test/alarm/alarm_test.cpp:24]: The former; I was relying on test order (I wasn't sure what best practice was w.r.t. order-independence; took a cue from NotAvailableAfterInit.) If we want tests to be order-independent I'll add an alarm before removing one.

File: controller/test/parameters/parameters_test.cpp (hunk 2 above)
Comment (SarahAdelaide) [reply controller/test/parameters/parameters_test.cpp:7]: You're right! These should all be one liners. Thanks :D

File: controller/test/alarm/alarm_test.cpp (hunk 1 above)
Comment (jlebar) [reply controller/test/alarm/alarm_test.cpp:24]: > If we want tests to be order-independent I'll add an alarm before removing one.

Yes, I think we want tests to be order-independent, otherwise modifying one test can break another.

To really make them hermetic we should reset the alarm state before each test in `SetUp`.

Comment (SarahAdelaide) [reply controller/test/alarm/alarm_test.cpp:21]: Okay! I've made the add and the remove test a single case so that they're order-independent. 

I'm not sure what you mean by "reset the alarm state before each test in SetUp". Do you mean for each test suite (by suite, I mean the collection of tests in a single file that are testing the same module) we should include an alarm reset in that suite's `SetUp`?

Comment (jlebar) [reply controller/test/alarm/alarm_test.cpp:21]: Ah, I think I understand.

A nonintuitive thing about gunit is that the test suite's `SetUp()` function is called before every test.  If you want to do one-time setup that's not repeated for every test, the function is `SetUpTestSuite()`.

> Do you mean for each test suite (by suite, I mean the collection of tests in a single file that are testing the same module) we should include an alarm reset in that suite's SetUp?

For instance, right now, `OverflowAlarmQueue ` leaves the alarm module with many alarms in its stack.  So then that affects the next test, which has to clear out the alarms.  It would be nice if we'd clear the alarm module's state before each test so that one test *can't* affect the next one.

It may be as simple as moving this code into `SetUp`:

```
  while (alarm_available()) {
    alarm_remove();
  }
```

File: controller/test/alarm/alarm_test.cpp [hunk 3]
```diff
@@ -12,18 +12,50 @@ class AlarmTest : public testing::Test {
...
+  dataID alarm_id_a = dataID::alarm_2;
+  alarm_add(alarm_id_a, alarm_data_a);
+  // checking that alarm data got copied, not referenced
+  memset(alarm_data_a, 'x', sizeof(alarm_data_a));
+  alarm_id_a = dataID::alarm_1;
+
+  enum dataID alarm_id_b;
+  uint32_t alarm_timestamp_b;
+  char alarm_data_b[ALARM_DATALEN];
+  int32_t res = alarm_read(&alarm_id_b, &alarm_timestamp_b, alarm_data_b);
```
Comment (jkff) [inline controller/test/alarm/alarm_test.cpp:51]: Drive-by comment: For functions that return errors, it'd be cool to use https://en.cppreference.com/w/cpp/language/attributes/nodiscard so that callers are required by the compiler to check the error.

File: controller/test/alarm/alarm_test.cpp (hunk 1 above)
Comment (SarahAdelaide) [reply controller/test/alarm/alarm_test.cpp:24]: oh! I see! I didn't realize setup was called before every test case. Thank you for the explanation. I've made the suggested change.

File: controller/test/alarm/alarm_test.cpp (hunk 3 above)
Comment (SarahAdelaide) [reply controller/test/alarm/alarm_test.cpp:50]: Thanks, Eugene! Maybe I'm misunderstanding the purpose of nodiscard, but it looks to me like it would be most usefully deployed in the codebase, not the test suites? Is this what you're thinking?

Comment (jkff) [reply controller/test/alarm/alarm_test.cpp:50]: Yes, I meant alarm_read itself (and other functions that may return an error), not its usage in this test.

File: controller/test/alarm/alarm_test.cpp [hunk 4]
```diff
@@ -6,24 +6,55 @@ class AlarmTest : public testing::Test {
 public:
   void SetUp() override {
     alarm_init();
+    while (alarm_available()) {
+      alarm_remove();
+    }
     // TODO: Reset Hal mock object.
```
Comment (jlebar) [inline controller/test/alarm/alarm_test.cpp:12]: The code we have now covers the TODO, I think.

**Comment by jlebar** [comment]
@SarahAdelaide this may have bitrotted some, but at a glance it looks like this is still worth landing!
